scrapy_po/utils.py

```diff
@@ -129,3 +133,19 @@ def build_instances(plan: andi.Plan, providers):
         else:
             instances[cls] = cls(**kwargs_spec.kwargs(instances))
     raise returnValue(instances)
+
+
+def callback_for(page_cls):
+    """Helper for defining callbacks for pages with to_item methods."""
+
+    # When the callback is used as an instance method of the spider, it expects
+    # to receive 'self' as its first argument. When used as a simple inline
+    # function, it expects to receive a response as its first argument.
+    #
+    # To avoid a TypeError, we need to receive a list of unnamed arguments and
+    # a dict of named arguments after our injectable.
+    def parse(*args, page: page_cls, **kwargs):
+        yield page.to_item()
+
+    setattr(parse, '__scrapy_po_callback', True)
```

Comment:
Since we cannot annotate the first argument with DummyRequest type (the first argument is now optional/dynamic), I'm using this private attribute to signal that this callback is a scrapy-po's callback and therefore it shouldn't be using response (see updated middleware source code).

Reply:
Couldn't it be a callable class and then check it's type?

Reply:
It could work like in the following example:

```python
class ItemCallback(Callable):

    pass


def callback_for(page_cls: Type[ItemPage]) -> Callable:
    """Helper for creating callbacks for ItemPage sub-classes."""
    if not issubclass(page_cls, ItemPage):
        raise TypeError(
            f'{page_cls.__name__} should be a sub-class of ItemPage.')

    if getattr(page_cls.to_item, '__isabstractmethod__', False):
        raise NotImplementedError(
            f'{page_cls.__name__} should implement to_item method.')

    class Callback(ItemCallback):

        # When the callback is used as an instance method of the spider,
        # it expects to receive 'self' as its first argument.
        # When used as a simple inline function,
        # it expects to receive a response as its first argument.
        #
        # To avoid a TypeError, we need to receive a list of
        # unnamed arguments and a dict of named arguments after our injectable.
        def __call__(*args, page: page_cls, **kwargs):  # type: ignore
            yield page.to_item()  # type: ignore

    return Callback()
```

And the check would be:

```python
def is_callback_using_response(callback: Callable):
    """Check whether the request's callback method is going to use response."""
    if issubclass(callback, ItemCallback):
        # The callback_for function was used to create this callback.
        return False

    ...
```

The issue is that Scrapy expects a function as a request's callback ([see here](https://docs.scrapy.org/en/latest/topics/request-response.html#scrapy.http.Request) and a Callable is an object. 

We would need to refactor [this line](https://github.com/scrapy/scrapy/pull/4500/files#diff-89dda711d8912841f3d91bd1816a6ca6R81) removing the `inspect.ismethod` predicate and update the logic to support class-based callables. Not to mention updating Scrapy's docs and making sure everything works as expected. 

I'm not sure which implications this might have.

Any thoughts? (cc @kmike)

Reply:
I think the current private attribute is a simple and straightforward check that works just fine.

Reply:
I was thinking it should be possible to implement parse as a class with `__call__` method and a descriptor (`__get__`), which ensures `self` is not passed (it would need https://github.com/scrapinghub/andi/pull/17 as well).

But given the issues you highlighted with Scrapy, a private attribute sounds like a good approach.